### PR TITLE
Refactor cache

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
refactoring

**Did you add tests for your changes?**
existing tests

**Does this PR introduce a breaking change?**
yes

**What needs to be documented once your changes are merged?**
* Modules are responsible for sorting of dependencies
* new hookable Cache class
* removed CachePlugin, added MemoryCachePlugin
* added Compiler.compilerPath
* removed cacheGroup argument from Compilation.addModule

### CODE REVIEW

File: test/watchCases/cache/child-compilation-cache/0/report-cache-counters-loader.js [hunk 1]
```diff
@@ -1,9 +1,26 @@
 var map = new Map();
 var currentWatchStepModule = require("../../../../helpers/currentWatchStep");
+var cacheMap = new WeakMap();
+
+const getCache = (assoicate, path) => {
```
Comment (montogeek) [inline test/watchCases/cache/child-compilation-cache/0/report-cache-counters-loader.js]: Just a typo `assoicate`

File: lib/FileSystemInfo.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+			parallelism: 2,
+			processor: this._readContextTimestamp.bind(this)
+		});
+	}
+
+	/**
+	 * @param {Map<string, FileSystemInfoEntry>} map timestamps
+	 * @returns {void}
+	 */
+	addFileTimestamps(map) {
```
Comment (sokra) [inline lib/FileSystemInfo.js]: This need to be fed by Watching

File: lib/NormalModule.js [hunk 3]
```diff
@@ -562,35 +567,55 @@ class NormalModule extends Module {
...
+									return callback(EARLY_RETURN_ERROR);
+								callback();
+							});
+						},
+						callback
+					),
+				callback =>
+					asyncLib.each(
+						this.buildInfo.contextDependencies,
+						(context, callback) => {
```
Comment (sokra) [inline lib/NormalModule.js]: I would be great to have a test for this.

File: lib/Compiler.js [hunk 4]
```diff
@@ -100,11 +100,14 @@ class Compiler {
...
+		this.root = this;
 		/** @type {string} */
 		this.outputPath = "";
 
 		this.outputFileSystem = null;
 		this.inputFileSystem = null;
+		this.watchFileSystem = null;
 
 		/** @type {string|null} */
 		this.recordsInputPath = null;
```
Comment (sokra) [inline lib/Compiler.js]: remove fileTimestamps and contextTimestamps properties

File: lib/Watching.js [hunk 5]
```diff
@@ -93,11 +103,13 @@ class Watching {
...
+			this.compiler.cache.beginIdle();
 			this.handler(err, stats);
 			return;
 		}
 
 		this.compiler.hooks.done.callAsync(stats, () => {
+			this.compiler.cache.beginIdle();
 			this.handler(null, stats);
 			if (!this.closed) {
 				this.watch(
```
Comment (sokra) [inline lib/Watching.js]: Replace this:

``` js
this.compiler.fileTimestamps = fileTimestamps;
this.compiler.contextTimestamps = contextTimestamps;
```

**Pull request merged**